Sources/SmithyCBOR/Writer/Writer.swift

```diff
@@ -0,0 +1,224 @@
+        self.nodeInfo = nodeInfo
+    }
+
+    public required init(nodeInfo: NodeInfo, parent: Writer? = nil) {
+        self.nodeInfo = nodeInfo
+        self.parent = parent
+    }
+
+    public func data() throws -> Data {
+        let encoder = try CBOREncoder()
+        try encode(encoder: encoder)
+        let encodedBytes = encoder.getEncoded()
+        return Data(encodedBytes)
+    }
+
+    private func encode(encoder: CBOREncoder) throws {
+        if let cborValue = self.cborValue {
+            // Encode the CBOR value directly if it exists
+            encoder.encode(cborValue)
+        } else if !children.isEmpty {
+            // Encode as an indefinite-length map
+            encoder.encode(.indef_map_start)
+            for child in children {
+                guard !(child.cborValue == nil && child.children.isEmpty) else {
+                    continue
+                }
+                encoder.encode(.text(child.nodeInfo)) // Key for the child
+
+                if let cborValue = child.cborValue {
+                    encoder.encode(cborValue) // Encode the value directly
+                } else {
+                    try child.encode(encoder: encoder) // Recursively encode nested maps/arrays
+                }
+            }
+            encoder.encode(.indef_break)
+        } else {
+            // No value and no children: encode an empty map
+            encoder.encode(.indef_map_start)
+            encoder.encode(.indef_break)
+        }
+    }
+
+    public subscript(nodeInfo: NodeInfo) -> Writer {
+        if let child = children.first(where: { $0.nodeInfo == nodeInfo }) {
+            return child
+        } else {
+            let newChild = Writer(nodeInfo: nodeInfo, parent: self)
+            children.append(newChild)
+            return newChild
+        }
+    }
+
+    public func write(_ value: Bool?) throws {
+        guard let value else { return }
+        self.cborValue = .bool(value)
+    }
+
+    public func write(_ value: String?) throws {
+        guard let value else { return }
+        self.cborValue = .text(value)
+    }
+
+    public func write(_ value: Float?) throws {
+        guard let value else { return }
+        self.cborValue = .double(Double(value))
+    }
+
+    public func write(_ value: Double?) throws {
+        guard let value else { return }
+        self.cborValue = .double(value)
+    }
+
+    public func write(_ value: Int?) throws {
+        guard let value else { return }
+        self.cborValue = .int(Int64(value))
+    }
+
+    public func write(_ value: Int8?) throws {
+        guard let value else { return }
+        self.cborValue = .int(Int64(value))
+    }
+
+    public func write(_ value: Int16?) throws {
+        guard let value else { return }
+        self.cborValue = .int(Int64(value))
+    }
+
+    public func write(_ value: UInt8?) throws {
+        guard let value else { return }
+        self.cborValue = .uint(UInt64(value))
+    }
```

Comment:
Should the integer types be written using "smallest" type that can fit the value, instead of Int64 / UInt64 here? I.e., SEP says `For example, a field modeled as a long with a value of 1 should be sent in a single byte.` so checking int size might be necessary to use minimum number of bits here

Reply:
Int64/UInt64 is what CRT supports so here we're writing a SmithyWriter supported swift value to a value that's ready to be CBOR encoded by CRT and then sent. CRT handles using the "smallest" type to fit the value, we handle moving a Swift type to a CBOR supported type. These write functions are not what gets sent. 

https://github.com/awslabs/aws-crt-swift/blob/697efa7317bc7110fd9b249722d7fcb37092e839/Source/AwsCommonRuntimeKit/crt/CBOR.swift#L9